windows-agent/internal/config/config.go

```diff
@@ -156,27 +158,86 @@ func (c *Config) SetUserLandscapeConfig(ctx context.Context, landscapeConfig str
 		return errors.New("attempted to set a user-provided landscape configuration when there already is a higher priority one")
 	}
 
+	landscapeConfig, err := completeLandscapeConfig(landscapeConfig, c.Landscape.UID)
+	if err != nil {
+		return fmt.Errorf("config: could not complete Landscape configuration: %v", err)
+	}
+
 	isNew, err := c.set(&c.Landscape.UserConfig, landscapeConfig)
 	if err != nil {
 		return errors.New("config: could not set Landscape configuration")
 	}
 
 	if isNew {
-		c.notifyLandsape(ctx, landscapeConfig, c.Landscape.UID)
+		c.notifyLandscape(ctx, landscapeConfig, c.Landscape.UID)
 	}
 
 	return nil
 }
 
-// SetLandscapeAgentUID overrides the Landscape agent UID.
-func (c *Config) SetLandscapeAgentUID(uid string) error {
-	if _, err := c.set(&c.Landscape.UID, uid); err != nil {
-		return fmt.Errorf("config: could not set Landscape agent UID: %v", err)
+// SetLandscapeAgentUID overrides the Landscape agent UID and notify listeners.
+func (c *Config) SetLandscapeAgentUID(ctx context.Context, uid string) error {
+	isUpdated, conf, err := c.setAgentUIDAndUpdateClientConf(ctx, uid)
+	if err != nil {
+		return err
+	}
+
+	if isUpdated {
+		log.Debugf(ctx, "config: notifying Landscape config listeners about agent UID change: %s", uid)
+		c.notifyLandscape(ctx, conf, uid)
 	}
 
 	return nil
 }
 
+// setAgentUIDAndUpdateClientConf sets the agent UID and updates the client configuration, returning the new configuration and the 'true' boolean flag to indicate that the configuration has changed,
+// allowing the caller to trigger notifications, preventing that from happening while we hold a lock.
+func (c *Config) setAgentUIDAndUpdateClientConf(ctx context.Context, uid string) (bool, string, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if err := c.load(); err != nil {
+		return false, "", fmt.Errorf("config: could not set Landscape agent UID: %v", err)
+	}
+
+	if c.Landscape.UID == uid {
+		log.Info(ctx, "config: no changes in the agent UID")
+		return false, "", nil
+	}
+
+	c.Landscape.UID = uid
+
+	// From this point on we have something to dump, but since `c.dump()` affects the values returned from this function, let's add yet one more indirection,
+	// setting the values we want to return first, and then deciding based on the result of `c.dump()`
+	mustNotify, updatedConf, err := func() (bool, string, error) {
```

Comment:
Setting `c.Landscape.*Config`  has the side-effect of changing what we dump to disk. In the paths we'd fail we don't touch those fields, thus dumping in those cases only writes the new agent UID. If dump itself fails, we don't want to trigger notifications, because listeners can get an inconsistent state if they call into the Config. Yet, by removing the boolean flag from the returned values as mentioned above, simplifies the code enough so the closure is no longer necessary.